app/models/user/email.rb

```diff
@@ -0,0 +1,36 @@
+# Represents an email address belonging to a user.
+class User::Email < ActiveRecord::Base
+  schema_validations except: :primary
+  validates :primary, inclusion: [true, false]
+  validates :primary, uniqueness: { scope: [:user_id], conditions: -> { where(primary: true) } }
+  validates :email, uniqueness: { case_sensitive: false }
+  validates :email, format: Devise.email_regexp
+
+  belongs_to :user, inverse_of: :emails
+
+  # Set the current email as primary.
+  #
+  # @return [Boolean] True if transaction was done successfully, otherwise nil.
+  def primary!
+    User::Email.transaction do
+      user.unset_primary_email
+      update_attributes!(primary: true)
+      true
+    end
+  end
+
+  # Properly destroy current email.
+  #
+  # @return [Boolean] True if the the email was destroyed, otherwise false.
+  def destroy!
```

Comment:
I directly called this function to destroy an email, call back won't work because destroy current email and set another email as primary should be in one transaction.